### ISSUE TITLE
VOLT / AMPERA support

### ISSUE THREAD

**Issue by liebefuroscar** [openvehicles/Open-Vehicle-Monitoring-System-3#196]
Opel Ampera / Chevrolet Volt users need to be included with priority:
- Preclimatization (OBD MX already has it)
- Load amperage selection

In a less priority way:
- The estimated autonomy is displayed (show 0)
- Tire pressure shown (show 0 psi)
- The motor temperature is displayed (show 0)
- Odometer Km shown in the app (show 0)

We are many users of Opel Ampera / Chevrolet Volt with OVMS that we miss these functions, which already has, for example, Tesla Roadster or Nissan Leaf.

I am also sure that if these features are added, many more users of these VE will buy an OVMS.
Thank you

![photo5917743255284461601](https://user-images.githubusercontent.com/25435711/51430147-1265d580-1c17-11e9-8de0-fe31300d3e2e.jpg)
![photo5917743255284461602](https://user-images.githubusercontent.com/25435711/51430148-1265d580-1c17-11e9-8c60-239241ede49f.jpg)

**Comment by raulcorreo** [comment]
Por favor, necesitamos estás funciones.
Hemos confiado comprando el dispositivo... Creo que lo merecemos....

**Comment by markwj** [comment]
We are limited by what the car can show on the CAN BUS, and what we can decode (or discover) as available. Very willing to implement these features, but we need the codes (which are not publicly available).

**Comment by lolvince** [comment]
Did you already see that ?
https://m.youtube.com/watch?v=xcfdSGDaqfs

**Comment by mjuhanne** [comment]
I'm in the middle of implementing extended Volt/Ampera support for OVMS (see my fork for preliminary work) and I'll make a pull request when it gets more stable and is usable on the mobile app.
However most of the interesting stuff and functionality happens on Single-Wire CAN (SWCAN, Low speed GMLAN) bus and unfortunately the CAN transceivers OVMS uses (MCP2515) won't be able to drive SWCAN with full functionality. It should be able to read and write CAN messages without bus errors, but waking the car (for unlocking/locking doors, remote heating, getting metrics updates etc) requires driving the SWCAN bus to 12 volts (High Voltage Wake Up) and this MCP2515 is unable to do. 
To fully utilize OVMS on Volt/Ampera, a proper SWCAN transceiver is needed (like TH8056), so I ended up designing an [add-on SWCAN board for OVMS](https://github.com/mjuhanne/OVMS-SWCAN).  At this stage everything is very experimental and you need some DIY/tinkering skills to produce the pcb, source the components and hand-solder the SMDs..

**Comment by SirSheik** [comment]
Good job mjuhanne 
In Spain there are many users who would be interested. Would it be possible to ask a company to mount the board?
Thx

**Comment by Artanicus** [comment]
Any updates here @mjuhanne ? Did you get anywhere with the hardware?

**Comment by mjuhanne** [comment]
@Artanicus 
Hi,

I haven't done any development or testing using OVMS for over year now since I sold my Ampera and bought a Tesla Model S, but I can comment on its status back then:

- The add-on SWCAN HW board works ok, but the OVMS firmware needs to be rebuilt with a special support for the board. 
- The board itself is needed if you need to wake up the car in order to poll the car status or the turn on the heater
- I remember that most of the useful CAN messages were deciphered and processed (outside/cabin temperature, alarm status)
-  Unfortunately the MCU in my car (Finnish MY2014) would not respond to Onstar CAN messages (commands to turn on heater/preconditioning,  lock/unlock doors etc). This may be because there was no Onstar module installed in the first place and most likely the main MCU was not configured to respond to those commands. The Onstar CAN messages OVMS sends are identical to the ones used by MyVoltStar, which I also used but with no luck. So I would assume that if you can use MyVoltStar to control your car with success, it can be done with OVMS+SWCAN board as well (but I haven't heard any success stories so far)
- I was able to turn on heater/preconditioning using diagnostic CAN commands sniffed from GDS2.  There's a configuration for using them OVMS for kind of backup solution, but I found the diagnostic commands to be quite finicky (heating would turn off automatically after a while and forcing the heating back again resulted in variable success). Lock/unlock doors using diagnostic commands worked too, but the alarm wasn't disabled so opening the doors would result in alarm going off :) I did not find a way to disable alarm using CAN commands.

In summary, if MyVoltStar can be used to control your car then OVMS+SWCAN board combo might be worth your while. Otherwise you are stuck with just status updates which are IMHO not that useful (In my case 95% of the occasions I use Tesla app is for climate control).

**Comment by RobTuckfield** [comment]
I've got a Volt with an OnStar module, if you need someone to test.

I'm a decent Python dev, and lately I've been trying to learn this car stuff, but I don't think I can contribute much to the code.

**Comment by melaz** [comment]
Hi community, is there any updates? Or something I can help? Recently bought Chevy volt 2013 and ordered ovms. How can I help with developing?

**Comment by markwj** [comment]
We are trying to produce a standard SWCAN board. See here:

https://github.com/openvehicles/Open-Vehicle-Monitoring-System-3/issues/704

Once that is produced, then perhaps a developer will step forward to see what can be done with it on the Volt/Ampera.

**Comment by TGC-TECH** [comment]
Any updates on this? I own a gen 1 (2014) Chevy Volt (US model) and while I don't yet have OVMS hardware, I could get a module with the SWCAN driver and could be willing to give this a go in terms of testing if there is still developer interest. 

I see Chevy Volt is listed as supported in the main readme but I also don't see any notes about it on docs.openvehicles.com

**Comment by dexterbg** [comment]
I can only tell from the source, that the current SWCAN integration can be used by the Volt adapter. Configure "External SWCAN adapter installed" in the Volt feature configuration, then reboot. This has been done by @kssmll Alexander Kiiashko, and has been included in the `edge` builds since end of February, but without any user feedback that I know of.

The user guide would need a maintainer who actually owns a Volt/Ampera. You're welcome to work on this.

**Comment by lululombard** [comment]
FYI, I'm currently adding support for Gen 2 Volts 

<img width="1174" height="453" alt="Image" src="https://github.com/user-attachments/assets/971f4e50-0e89-46a8-af13-8b902b061f9f" />

<img width="1166" height="655" alt="Image" src="https://github.com/user-attachments/assets/a2931f2b-ac64-433e-8298-becb156d8de1" />

I'm even adding a charge limit which is not even a standard feature, by deferring charging forever

<img width="1159" height="637" alt="Image" src="https://github.com/user-attachments/assets/b5b0e1ad-66c8-49a4-911f-1fdfe126b603" />

Once ready I'll make my fork public and open a PR

**Comment by liebefuroscar** [comment]
Hi everyone,

I wanted to provide some feedback as an Opel Ampera Gen 1 owner.

I'm currently running:

OVMS firmware: 3.3.006
Vehicle: Opel Ampera Gen 1
External SWCAN adapter: Installed and enabled

I've tested some of the SWCAN functionality:

✅ Lock / Unlock works correctly.
⚠️ Preconditioning does not complete successfully.

When I start preconditioning from OVMS, the car turns on the exterior lights and the HVAC blower starts, but only for a few seconds before stopping.

Interestingly, if I then unlock and open the driver's door, it seems as if the car resumes the blower operation that had been interrupted. It almost feels like the preconditioning sequence starts but never fully transitions into the normal heating mode while the car remains locked.

If it helps, I'd be happy to test experimental firmware, collect logs, or provide any additional information needed. I have both a Gen 1 Ampera and the external SWCAN adapter available for testing.

Thanks to everyone who continues working on Volt/Ampera support.
